src/Framework/Constraint/Connective.php

```diff
@@ -0,0 +1,113 @@
+<?php declare(strict_types=1);
+/*
+ * This file is part of PHPUnit.
+ *
+ * (c) Sebastian Bergmann <sebastian@phpunit.de>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+namespace PHPUnit\Framework\Constraint;
+
+/**
+ * Abstract base class for connectives (n-ary) logical operators.
+ *
+ * Infix notation of the operator is assumed.
+ */
+abstract class Connective extends Operator
```

Comment:
Not so sure about the naming and the inheritance here, perhaps something along the lines of `CompositeConstraint` would be better?

What do you think?

Reply:
``CompositeConstraint`` does not reveal the actual purpose of these objects (it only emphasizes composition). The origin of the ``Connective`` class name is [Logical Connective](https://en.wikipedia.org/wiki/Logical_connective), which denotes a logical operator (binary or unary, Wikipedia says). I adopted this name for operators that accept arbitrary number of arguments (0 to inf). We can still look for a better name.

Reply:
Then `LogicalConnective` is probably the name we're looking for. After all, we already have `LogicalAnd`, etc.

Reply:
With new commits, I've renamed base classes to the following: ``Unary`` -> ``UnaryOperator``, ``Connective`` -> ``BinaryOperator``. I'll set this as resolved for now, but we can get back to it and rename these classes further.